orderer/common/cluster/block_fetcher.go

```diff
@@ -0,0 +1,489 @@
+	}
+}
+
+type impatientStreamCreator func(in *common.Envelope) (*impatientAttestationStream, error)
+
+// impatientAttestationStream aborts the stream if it waits for too long for a message.
+type impatientAttestationStream struct {
+	waitTimeout time.Duration
+	orderer.BlockAttestations_BlockAttestationsClient
+	cancelFunc func()
+}
+
+func (stream *impatientAttestationStream) abort() {
+	stream.cancelFunc()
+}
+
+// Recv blocks until a response is received from the stream or the
+// timeout expires.
+func (stream *impatientAttestationStream) Recv() (*ab.BlockAttestationResponse, error) {
+
+	// Initialize a timeout to cancel the stream when it expires
+	timeout := time.NewTimer(stream.waitTimeout)
+	defer timeout.Stop()
+
+	responseChan := make(chan attestationErrorAndResponse, 1)
+
+	// receive waitGroup ensures the goroutine below exits before
+	// this function exits.
+	var receive sync.WaitGroup
+	receive.Add(1)
+	defer receive.Wait()
+
+	go func() {
+		defer receive.Done()
+		resp, err := stream.BlockAttestations_BlockAttestationsClient.Recv()
+		responseChan <- attestationErrorAndResponse{err: err, resp: resp}
+	}()
+
+	select {
+	case <-timeout.C:
+		stream.cancelFunc()
+		return nil, errors.Errorf("didn't receive a response within %v", stream.waitTimeout)
+	case respAndErr := <-responseChan:
+		return respAndErr.resp, respAndErr.err
+	}
+}
+
+// newImpatientStream returns a ImpatientStreamCreator that creates impatientStreams.
+func newImpatientStream(conn *grpc.ClientConn, waitTimeout time.Duration) impatientStreamCreator {
+	return func(in *common.Envelope) (*impatientAttestationStream, error) {
+		abc := orderer.NewBlockAttestationsClient(conn)
+		ctx, cancel := context.WithCancel(context.Background())
+
+		stream, err := abc.BlockAttestations(ctx, in)
+		if err != nil {
+			cancel()
+			return nil, err
+		}
+
+		once := &sync.Once{}
+		return &impatientAttestationStream{
+			waitTimeout: waitTimeout,
+			// The stream might be canceled while Close() is being called, but also
+			// while a timeout expires, so ensure it's only called once.
+			cancelFunc: func() {
+				once.Do(cancel)
+			},
+			BlockAttestations_BlockAttestationsClient: stream,
+		}, nil
+	}
+}
+
+func (bf *AttestationPuller) PullAttestations(seq uint64) []*orderer.BlockAttestation {
+	// endpointsToProbe
+	endpointsToProbe := bf.endpoints
+	// Pull Attestations from the endpoints
+	// add a Sync.WaitGroup and probe using go routines
+	waitGrp := sync.WaitGroup{}
+	// recieve the blocks using channels to a slice
+
+	// verify the slice and decide whether current node had byzantine behavior
+	type bps struct {
+		block *orderer.BlockAttestation
+		ec    EndpointCriteria
+	}
+
+	bpsInfo := make(chan *bps, len(endpointsToProbe))
+
+	waitGrp.Add(len(endpointsToProbe))
+	// start probing
+	for _, ec := range endpointsToProbe {
+		// get a puller
+
+		go func(ec EndpointCriteria) {
+			defer waitGrp.Done()
+			block, _ := bf.pullAttestationBlock(ec, seq)
```

Comment:
What will be the behaviour if pullAttestationBlock returns with err?   can we add a log statement if it fails to pull attestation for an endpoint.

Reply:
Added a log. It should return a nil block.